extensions/amp-addthis/0.1/amp-addthis.js

```diff
@@ -0,0 +1,190 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Embeds an AddThis widget.
+ * The data-pubId and data-widgetId can be found easily in the AddThis dashboard at addthis.com.
+ * Example:
+ * <code>
+ * <amp-addthis
+ *   width="320"
+ *   height="92"
+ *   layout="responsive"
+ *   data-pubId="ra-59c2c366435ef478"
+ *   data-widgetId="0fyg">
+ * </amp-addthis>
+ * </code>
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {setStyles} from '../../../src/style';
+import {removeElement} from '../../../src/dom';
+import {user} from '../../../src/log';
+
+import {ConfigManager} from './ConfigManager';
+import {WIDGET_ID_PROP, ORIGIN, ICON_SIZE, ALT_TEXT} from './constants';
+
+// This `configManager` will be shared by all AmpAddThis elements on a page, to prevent unnecessary
+// HTTP requests, get accurate analytics, etc.
+const configManager = new ConfigManager();
+
+export function getConfigManager() {
+  return configManager;
+}
+
+class AmpAddThis extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?HTMLIFrameElement} */
+    this.iframe_ = null;
+
+    /** @private {?Promise} */
+    this.iframeLoadPromise_ = null;
+
+    /** @private {string} */
+    this.pubId_ = '';
+
+    /** @private {string} */
+    this.widgetId_ = '';
+  }
+
+  /**
+   * @param {boolean=} opt_onLayout
+   * @override
+   */
+  preconnectCallback(opt_onLayout) {
+    const {preconnect} = this;
+    preconnect.url(ORIGIN, opt_onLayout);
+    preconnect.url('https://m.addthis.com', opt_onLayout);
+    preconnect.url('https://m.addthisedge.com', opt_onLayout);
+    preconnect.url('https://cache.addthiscdn.com', opt_onLayout);
+  }
+
+  /**
+   * @return {boolean}
+   * @override
+   */
+  renderOutsideViewport() {
+    return true;
+  }
+
+  /**
+   * Note: Does no actual building (building is deferred until layout). Simply checks for/sets
+   * required attributes (pudId/widgetId).
+   * @override
+   */
+  buildCallback() {
+    const {element} = this;
+    const pubId = element.getAttribute('data-pubId') ||
+        element.getAttribute('pubId');
+    const widgetId = element.getAttribute('data-widgetId') ||
+        element.getAttribute('widgetId');
+
+    this.pubId_ = user().assert(
+        pubId,
+        'The data-pubId attribute is required for <amp-addthis> %s',
+        element
+    );
+    this.widgetId_ = user().assert(
+        widgetId,
+        'The data-widgetId attribute is required for <amp-addthis> %s',
+        element
+    );
+  }
+
+  createPlaceholderCallback() {
+    const {win} = this;
+
+    const placeholder = win.document.createElement('div');
+    placeholder.setAttribute('placeholder', '');
+    setStyles(placeholder, {
+      'background-color': '#eee',
+    });
+
+    const image = win.document.createElement('amp-img');
```

Comment:
Nit: we have a [createElementWithAttributes](https://github.com/ampproject/amphtml/blob/3b60020ca0478d83289458c2af7b92da8b8918db/src/dom.js#L162-L172) that could clean this up.

Reply:
Oh, nice! Thanks; I'll update the diff as soon as I can.